File: src/main/java/io/github/willslaunwhite/linkedlist/problems/ReverseLinkedList.java
package io.github.willslaunwhite.linkedlist.problems;

import io.github.willslaunwhite.linkedlist.ListNode;

/*
 * Given the head of a singly linked list, reverse the list, and return the reversed list.
 * 
 * Input: head = [1,2,3,4,5]
 * Output: [5,4,3,2,1]
 * 
 * Input: head = [1,2]
 * Output: [2,1]
 * 
 * Input: head = []
 * Output: []
 */

public class ReverseLinkedList {
    public ListNode reverseLinkedListIterative(ListNode head) {
        ListNode prev = null;
        while (head != null) {
            ListNode next = head.next;
            head.next = prev;
            prev = head;
            head = next;
        }

        return prev;
    }

    public ListNode reverseLinkedListRecursive(ListNode head) {
        return this.reverse(head, null);
    }

    public ListNode reverse(ListNode current, ListNode prev) {
        if (current == null) {
            return prev;
        }
        ListNode next = current.next;
        current.next = prev;
        return this.reverse(next, current);
    }
}

File: python/linked_list/problems/reverse_linked_list.py
from typing import Optional
from linked_list.list_node import ListNode


def reverse_linked_list_iterative(head: Optional[ListNode]) -> Optional[ListNode]:
    """
    Reverse the input linked list recursively.

    Args:
        head (ListNode): the head of the linked list.

    Returns:
        ListNode: the head of the reversed list.

    """

    prev: Optional[ListNode] = None
    while head != None:
        next: Optional[ListNode] = head.next
        head.next = prev
        prev = head
        head = next
    return prev


def reverse(
    current: Optional[ListNode], prev: Optional[ListNode] = None
) -> Optional[ListNode]:
    """
    Reverse the input linked list iteratively

    Args:
        head (Optional[ListNode]): the head of the linked list.

    Returns:
        Optional[ListNode]: the head of the reversed list.
    """
    if not current:
        return prev

    next: Optional[ListNode] = current.next
    current.next = prev
    return reverse(next, current)


def reverse_linked_list_recursive(head: Optional[ListNode]) -> Optional[ListNode]:
    """
    calls the method to reverse the linked list recursively

    Args:
        head (Optional[ListNode]): the head of the linked list

    Returns:
        Optional[ListNode]: the head of the reversed
    """
    return reverse(head)
